### ISSUE TITLE
Design feedback

### ISSUE THREAD

**Issue by mathstuf** [d-e-s-o/http-endpoint#1]
i, I was pointed here by [`@michael_j_ward` on Reddit](https://old.reddit.com/user/michael_j_ward) in response to [my submission](https://old.reddit.com/r/rust/comments/np41l2/designing_rust_bindings_for_rest_apis/) about REST API bindings design in Rust. I know this is unsolicited advice, but I thought I'd drop in to share some observations.

I looked a bit at the API exposed by the crate here and I have a few observations:

  - `base_url` should be purely within the client's purview rather than anything to do with the endpoint itself
  - the `Output` type should be free rather than tied to a given `Endpoint`; a specific deserialization error case should be handled in case there is a problem
  - there's no way to apply combinators to endpoints to modify their behavior (e.g., in the design I describe in my post, pagination is a modifier on an endpoint)
  - I'd recommend using a parameterized `Error` type which holds the conversion and `ApiError` types rather than forcing that to be done on the endpoint side as well

Looks like a decent start to a generic REST API representation at least.

**Comment by d-e-s-o** [comment]
Thanks for your comment/advice! So the crate started off just as a way to satisfy my own needs. That's why at least until `0.5` the design may not have been the cleanest -- I wanted to push a bunch of stuff into a generic layer while minimizing client specific code. I'd argue that's also why the documentation is rather...sparse ;-)

In general I'd say it's rather hard to generically and perhaps declaratively describe a REST endpoint -- in many ways it's too much of a free-for-all with mostly conventions instead of hard rules on the endpoint implementer's side. I tried to minimize reliance on such rules, but won't claim freedom from them.

To your points:
> * `base_url` should be purely within the client's purview rather than anything to do with the endpoint itself

Conceptually, yes. I used it as a way to overwrite the default, though. Consider the case where you want to talk to ten endpoints, with 9 of them at `api.foobar.com` and another one at `data.foobar.com`. With a generic client it gets hard to cater to cases like this, and unfortunately we can't rule them out. See for instance https://github.com/d-e-s-o/apca/blob/ef3a42aa7d6c39f70c4c54259efd5ffc49bdbafe/src/data/v1/bars.rs#L134 for an override of the generic https://github.com/d-e-s-o/apca/blob/ef3a42aa7d6c39f70c4c54259efd5ffc49bdbafe/src/client.rs#L116

I have considered splitting functionality more and factoring out more generic bits and pieces in the past (which may alleviate this short coming), but am yet to find a coherent design.

> * the `Output` type should be free rather than tied to a given `Endpoint`; a specific deserialization error case should be handled in case there is a problem

I believe to only have understood what you are saying after reading your post. Based on my understanding of it, my assumptions and usage scenarios seem different. In my view, every endpoint returns some kind of data (some only `()`), but what is being returned is different on a per-endpoint basis. So it *is* the endpoint that defines what that data is. I can see the argument you brought up, I am just not sure I agree with the solution.

> * there's no way to apply combinators to endpoints to modify their behavior (e.g., in the design I describe in my post, pagination is a modifier on an endpoint)

I consider pagination as something built on top (as there are many ways to go about it), e.g., https://github.com/d-e-s-o/apca/blob/ef3a42aa7d6c39f70c4c54259efd5ffc49bdbafe/src/api/v2/account_activities.rs#L621. Right now I don't have excessive usage of repeating patterns and as such haven't thought much about a combinator style approach and whether/how it would fit in. But it does sound interesting.

> * I'd recommend using a parameterized `Error` type which holds the conversion and `ApiError` types rather than forcing that to be done on the endpoint side as well

So yes, that is almost how it is meant to be (if I understand your point correctly). But I still need/want the API error and the conversion error. One of the main aspects of the `EndpointDef` macro is the automatic generation of error variants. If you want that you need a dedicated type just for that (can't be user provided). Furthermore, I added hooks for parsing/deserialization/conversion stuff into the macro because I wanted to be able to easily provide a general implementation, but at the same time allow for it to be overwritten if need be. So that's why I have a a `ConversionError` in there. It is user-specified because it may actually be an enumeration of different error types. E.g., I club formurlencode errors and JSON errors in it.

I hope that explains some of the design decisions.

**Comment by mathstuf** [comment]
> So it is the endpoint that defines what that data is. I can see the argument you brought up, I am just not sure I agree with the solution.

If you're building the service and the client together, sure. But not all clients care about all of the data. Saving the time to parse JSON strings into dates if all I care about is some settings flags is still very useful.

> I consider pagination as something built on top (as there are many ways to go about it)

Right. That's why there is the combinator pattern. The `Pageable` trait would have to probably learn to indicate which pagination strategy to use for the multitude of cases.

> But I still need/want the API error and the conversion error.

Yes. They're still available in `ApiError<Client::Error>`, but just "hidden" behind known variants.

I suspect that we're approaching these things from very different angles. I myself never really design REST API implementations and only use them as a client, but it sounds like this is intended to also be useful for the server too?

**Comment by d-e-s-o** [comment]
> > So it is the endpoint that defines what that data is. I can see the argument you brought up, I am just not sure I agree with the solution.
> 
> If you're building the service and the client together, sure. But not all clients care about all of the data. Saving the time to parse JSON strings into dates if all I care about is some settings flags is still very useful.

Yes, I agree. The main use case for me was to have sensible defaults available.

> > I consider pagination as something built on top (as there are many ways to go about it)
> 
> Right. That's why there is the combinator pattern. The `Pageable` trait would have to probably learn to indicate which pagination strategy to use for the multitude of cases.

:+1: 

> > But I still need/want the API error and the conversion error.
> 
> Yes. They're still available in `ApiError<Client::Error>`, but just "hidden" behind known variants.

Hm, I didn't see much mention of them. Are they "generic" variants in the `ApiError` or `Client::Error`? They are the main reason I have to use a macro: I treat them as endpoint specific and so I "generate" a specific error for each endpoint.

> I suspect that we're approaching these things from very different angles. I myself never really design REST API implementations and only use them as a client, but it sounds like this is intended to also be useful for the server too?

Yes, it's also my take away; probably because while trying to stay generic we both had specific use cases that we wanted to cater to. But no, I am looking at the client side only.
